### ISSUE TITLE
DataFrame.apply returns Series of type object instead of correct dtype

### ISSUE THREAD

**Issue by jrbrodie77** [pandas-dev/pandas#26223]
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
df = pd.DataFrame([[1.0, 2.0, 3.0, 'A'],
                   [1.0, 2.0, 2.0, 'A'],
                   [2.0, 4.3, 2.3, 'B'],
                   [6.5, 3.4, 2.1, 'B']],
                   columns=['n0', 'n1', 'n2', 'letter'])

def print_series_dtype(series):
    print(series.name, series.dtype)
    
def print_dataframe_dtypes(dataframe):
    print("DF dtypes:", list(dataframe.dtypes))
    print("DF colnames", list(dataframe.columns))
    dataframe.apply(print_series_dtype)

df.groupby('letter').apply(print_dataframe_dtypes)

# Output
DF dtypes: [dtype('float64'), dtype('float64'), dtype('float64'), dtype('O')]
DF colnames ['n0', 'n1', 'n2', 'letter']
n0 object
n1 object
n2 object
letter object


```
#### Problem description
DataFrame.gropby(..).apply(<func>) provides func with a series that has a dtype of object, regardless of the original dtype.

The code snippet above shows that the series dtype is mismatched. 

#### Expected Output
I would expect the dtypes to propagate unchanged. (in this case, it caused some downstream code to seriously misbehave).

#### Output of ``pd.show_versions()``
In [24]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.4
pytest: 3.8.2
pip: 19.0.3
setuptools: 40.2.0
Cython: 0.29.7
numpy: 1.15.2
scipy: 1.1.0
pyarrow: 0.11.1
xarray: None
IPython: 7.3.0
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 3.0.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

**Comment by WillAyd** [comment]
apply is looking for some kind of return value to infer the type from, which printing does not provide (None implicitly gets returned, hence the object inference)

You are looking for df.dtypes

**Comment by WillAyd** [comment]
Hmm sorry got mixed up with the groupby in there. The example can be simplified a bit.

```python
In [8]: def get_dtype(ser):
   ...:     return ser.dtype
   ...:

In [9]: df.apply(get_dtype)
Out[9]:
n0        object
n1        object
n2        object
letter    object
dtype: object

In [10]: df.drop(columns=['letter']).apply(get_dtype)
Out[10]:
n0    float64
n1    float64
n2    float64
dtype: object
```

So it looks like inclusion of the `letter` column might be coercing the others to object before the apply. 

There might already be an open issue for this if you don't mind checking for duplicates

**Comment by simonjayhawkins** [comment]
> I would expect the dtypes to propagate unchanged. (in this case, it caused some downstream code to seriously misbehave).

for DataFrame.groupy where objects passed to the function are Series objects whose index is the DataFrame’s index (axis=0), this would perhaps be the expected behaviour. However, DataFrame.groupy also allows grouping with axis=1.

the code sample is maybe looking at just part of the internals, that are not relevant when the DataFrame is reconstructed from the returned series...

```python
Python 3.7.3 (default, Mar 27 2019, 17:13:21) [MSC v.1915 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pandas as pd
>>> pd.__version__
'0.25.0.dev0+455.gd74901b1a'
>>> df = pd.DataFrame([[1.0, 2.0, 3.0, 'A'],
...                    [1.0, 2.0, 2.0, 'A'],
...                    [2.0, 4.3, 2.3, 'B'],
...                    [6.5, 3.4, 2.1, 'B']],
...                    columns=['n0', 'n1', 'n2', 'letter'])
>>> def get_dtype(ser):
...     print(ser.name, ser.dtype)
...     return ser
...
>>> df.apply(get_dtype).info()
n0 object
n0 float64
n1 float64
n2 float64
letter object
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 4 columns):
n0        4 non-null float64
n1        4 non-null float64
n2        4 non-null float64
letter    4 non-null object
dtypes: float64(3), object(1)
memory usage: 208.0+ bytes
>>>

```

is it possible to provide a minimal reproducible example that shows how downstream code misbehaves?

**Comment by jrbrodie77** [comment]
So I've poked around that this one again.  The behavior is definitely related to .apply running code more than once to sniff the types.   The full example is pasted below.  For some reason, if I raise an exception with assert when the code does the wrong thing, some assumptions are made and the code does the right thing!  (run with the assert line in or out and you'll see what I mean).

```
import numpy as np
import pandas as pd
df = pd.DataFrame([[1.0, 2.0, 3.0, 'A'],
                   [1.0, 2.0, 2.0, 'A'],
                   [2.0, 4.3, 2.3, 'B'],
                   [6.5, 3.4, 2.1, 'B']],
                   columns=['n0', 'n1', 'n2', 'letter'])

def groupby_funker(func):
    """Returns a function that applies func to a dataframe"""
    def col_func(col_series):
        if np.issubdtype(col_series.dtype, np.number):
            # Numeric column, apply func
            return func(col_series)
        else:
            # assert len(col_series.unique()) == 1   # Uncomment this line to fix the code!
            return col_series.iloc[0]

    def df_func(dataframe):
        # reduce=False stops the df from changing type to object.
        return dataframe.apply(col_func, reduce=False)
    
    return df_func


dataframe_mean = groupby_funker(np.mean)
result_df = df.groupby('letter').apply(dataframe_mean)

if np.isclose(result_df.loc['B', 'n1'],
              np.mean(df.loc[df.letter=='B', 'n1'])):
    print("\nMean function worked.")
else:
    print("\nMean function didn't work.")

result_df
```

**Comment by mroeschke** [comment]
I _think_ this is solved on master. Could use a test I suppose.

```
In [141]: In [8]: def get_dtype(ser):
     ...:    ...:     return ser.dtype
     ...:

In [142]:

In [142]: In [9]: df.apply(get_dtype)
     ...:
Out[142]:
n0        float64
n1        float64
n2        float64
letter     object
dtype: object

In [143]: In [10]: df.drop(columns=['letter']).apply(get_dtype)
     ...:
Out[143]:
n0    float64
n1    float64
n2    float64
dtype: object
```

**Comment by gabrielNT** [comment]
I think this is the same as #31466.

**Comment by tnwei** [comment]
Yup this issue is solved in #31466, think we can close this.

**Comment by mroeschke** [comment]
Yes, thanks!
